### PR TITLE
mul(dense, csr), mul(csr, dense) via sparse_mask_csr

### PR DESCRIPTION
**Pull request by cpuhrsch**
This adds basic coverage, but can be easily made more efficient by providing a native implementation.

Follow up work includes supporting CSR gradients for strided Tensors.

### CODE REVIEW

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp [hunk 1]
```diff
@@ -709,8 +709,12 @@ Tensor& mul_sparse_(Tensor& self, const Tensor& other) {
 
 Tensor& mul_out_sparse_csr(const Tensor& t_, const Tensor& src_, Tensor& r) {
   // // TODO: Use a specialized CSR kernel for performance if needed
-  TORCH_CHECK(t_.is_sparse_csr() || (t_.layout() == c10::kStrided && t_.dim() == 0), "mul(dense, sparse_csr) is not supported");
-  TORCH_CHECK(src_.is_sparse_csr() || (src_.layout() == c10::kStrided && src_.dim() == 0), "mul(sparse_csr, dense) is not supported");
+  if (t_.is_sparse_csr() && src_.layout() == kStrided) {
+    return mul_out_sparse_csr(t_, src_.sparse_mask(t_), r);
+  }
```
[the remark this answers is not in the document]
Comment (cpuhrsch) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp:714]: Currently we convert everything to COO for multiplication.

[the remark this answers is not in the document]
Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp:714]: If needed, we can improve that dispatch as well speaking of the CPU version.

File: test/test_sparse_csr.py [hunk 2]
```diff
@@ -1307,34 +1307,62 @@ def _test_spadd_shape(nnz, shape):
...
-            res = fn(y, x)
+            # Forward comparison
+            res_sparse_sparse = fn(y, x)
+            res_dense_sparse = fn(y.to_dense(), x)
+            res_sparse_dense = fn(y, x.to_dense())
             expected = fn(y.to_dense(), x.to_dense()).to_sparse_csr()
-            self.assertEqual(res, expected)
+            self.assertEqual(res_sparse_sparse, expected)
+            self.assertEqual(res_dense_sparse.to_dense(), expected.to_dense())
+            self.assertEqual(res_sparse_dense.to_dense(), expected.to_dense())
```
Comment (nikitaved) [inline test/test_sparse_csr.py:1322]: Why is there `to_dense` conversion?

Comment (cpuhrsch) [reply test/test_sparse_csr.py:1322]: While the result of mul(dense, csr) is csr, the result isn't fully compressed. To achieve that we'd need to remove all zeros from the output. When converting dense to csr via sparse_mask it's not also removing all the zeros, it's just applying the indices of the given argument.

Comment (nikitaved) [reply test/test_sparse_csr.py:1322]: Got it, I did not know that about `sparse_mask`. Maybe a comment could be of help.

**Review by nikitaved — APPROVED** [review]
LGTM! Thank you, @cpuhrsch !

**Review by mikaylagawarecki — APPROVED** [review]

**Pull request closed without merging**